components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java

```diff
@@ -2106,4 +2272,173 @@ private void handleInternalTokenRevocation(String consumerKey, Properties proper
+     * @throws IdentityOAuthClientException
+     */
+    private void filterAllowedFAPITokenAuthMethods(String authenticationMethod) throws IdentityOAuthClientException {
+
+        List<String> allowedAuthMethods = IdentityUtil.getPropertyAsList(
+                OAuthConstants.FAPI_CLIENT_AUTH_METHOD_CONFIGURATION);
+        if (authenticationMethod != null && !allowedAuthMethods.contains(authenticationMethod)) {
+            throw handleClientError(INVALID_REQUEST, "Invalid token endpoint authentication method requested.");
+        }
+    }
+
+    /**
+     * FAPI validation to restrict the signature algorithms.
+     * Link - https://openid.net/specs/openid-financial-api-part-2-1_0.html#algorithm-considerations
+     * @param signatureAlgorithm signature algorithm used to sign the assertions.
+     * @throws IdentityOAuthClientException
+     */
+    private void filterAllowedFAPISignatureAlgorithms(String signatureAlgorithm)
+            throws IdentityOAuthClientException {
+
+        List<String> allowedSignatureAlgorithms = IdentityUtil
+                .getPropertyAsList(OAuthConstants.FAPI_SIGNATURE_ALGORITHM_CONFIGURATION);
+        if (signatureAlgorithm != null && !allowedSignatureAlgorithms.contains(signatureAlgorithm)) {
+            throw handleClientError(INVALID_REQUEST, "Invalid signature algorithm requested");
+        }
+    }
+
+    /**
+     * FAPI validation to restrict the encryption algorithms.
+     * Link - https://openid.net/specs/openid-financial-api-part-2-1_0.html#encryption-algorithm-considerations
+     * @param encryptionAlgorithm
+     * @throws IdentityOAuthClientException
+     */
+    private void filterAllowedFAPIEncryptionAlgorithms(String encryptionAlgorithm)
+            throws IdentityOAuthClientException {
+
+        if (encryptionAlgorithm.equals(OAuthConstants.RESTRICTED_ENCRYPTION_ALGORITHM)) {
+            throw handleClientError(INVALID_REQUEST, "Invalid encryption algorithm requested");
+        }
+    }
+
+
+    /**
+     * If there are multiple hostnames in the registered redirect_uris,
+     * the Client MUST register a sector_identifier_uri.
+     * https://openid.net/specs/openid-connect-core-1_0.html#PairwiseAlg
+     * @param redirectURIs list of callback urls sent in the request
+     * @throws IdentityOAuthClientException
+     */
+    private void validateRedirectURIForPPID(List<String> redirectURIs, String sectorIdentifierURI)
+            throws IdentityOAuthClientException {
+
+        if (StringUtils.isEmpty(sectorIdentifierURI)) {
+            if (redirectURIs.size() > 1) {
+                String hostname = URI.create(redirectURIs.get(0)).getHost();
+                for (String redirectURI : redirectURIs) {
+                    URI uri = URI.create(redirectURI);
+                    if (!uri.getHost().equals(hostname)) {
+                        throw handleClientError(INVALID_REQUEST,
+                                "Sector identifier URI is needed for PPID calculation");
```

Comment:
Shall we mention the exact case for the error. its either callback urls with the same host or sector identifier uri being required.

Reply:
if the call bakc urls have multiple hosts, then the sector identifier uri should be sent in the request..